src/core/app.lua

```diff
@@ -124,6 +128,65 @@ function configure (new_config)
    counter.add(configs)
 end
 
+-- Claims a name for a program so it can be identified by name by other processes.
+--
+-- The name given to the function must be unique, if a name has been used before by
+-- an active process the function will error displaying an appropriate error message.
+-- Any other process can enumerate programs by their name and PID using enumerate_named_programs.
+function claim_name(name)
+   configuration[name] = name
+   local namedir = "by-name/" .. name
+   local namedirqualified = named_program_root .. "/" .. name
+   local piddir = shm.root .. "/" .. S.getpid()
+
+   -- Verify that the by-name directory exists.
+   shm.mkdir(namedir)
+   
+   -- Check if a symlink to a named program with the same name exists, if not, unlink it.
+   if S.lstat(namedirqualified) then
+      -- It exists, lets check what the PID is.
+      local piddir = S.readlink(namedirqualified)
+      local ps, pe = string.find(piddir, "/[^/]*$")
+      local pid = tonumber(string.sub(piddir, ps+1, pe))
+      if S.kill(pid, 0) then
+         -- It's a running process, we should display an error message.
+         error("Name has been claimed by a currently running process: "..namedirqualified)
```

Comment:
Is the kill necessary?  I guess we could grovel in /proc.  One question: are you sure that the process with this PID is the one that made the name dir?  Does the supervisor process guarantee that?  If not we could write the process start time into the per-pid directory or something.  Maybe this is not important.